File: wagerproof-android-native/app/src/main/java/com/wagerproof/app/features/gamecards/BestBookChip.kt
package com.wagerproof.app.features.gamecards

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.wagerproof.core.design.tokens.AppColors
import com.wagerproof.core.models.SportsbookMarketQuotes
import com.wagerproof.core.models.SportsbookQuote
import java.time.Duration
import java.time.Instant

/**
 * Port of iOS `BestBookChip` / `SportsbookLinesSheet`.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun BestBookChip(
    quotes: SportsbookMarketQuotes,
    selectedBookKeys: Set<String>,
    marketTitle: String,
    selectionTitle: String,
    formatLine: (Double) -> String,
    modifier: Modifier = Modifier,
) {
    val shown = quotes.preferred(selectedBookKeys) ?: return
    val fromSelection = quotes.isFromSelection(selectedBookKeys)
    var showingBoard by remember { mutableStateOf(false) }
    val label = when {
        selectedBookKeys.isEmpty() -> "BEST LINE"
        fromSelection && selectedBookKeys.size == 1 -> "YOUR BOOK"
        fromSelection -> "YOUR BOOKS"
        else -> "NOT AT YOUR BOOKS"
    }

    Row(
        modifier
            .clip(CircleShape)
            .background(AppColors.appSurfaceMuted.copy(alpha = 0.65f))
            .border(0.5.dp, AppColors.appBorder.copy(alpha = 0.45f), CircleShape)
            .clickable { showingBoard = true }
            .padding(horizontal = 10.dp, vertical = 7.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(7.dp),
    ) {
        SportsbookLogoView(shown.bookKey, null, SportsbookLogoStyle.COMPACT)
        Column(verticalArrangement = Arrangement.spacedBy(1.dp)) {
            Text(label, color = AppColors.appTextMuted, fontSize = 8.sp, fontWeight = FontWeight.Black)
            Text(
                priceLabel(shown, formatLine),
                color = AppColors.appTextPrimary,
                fontSize = 13.sp,
                fontWeight = FontWeight.Black,
            )
        }
        Text("▾", color = AppColors.appTextMuted, fontSize = 9.sp, fontWeight = FontWeight.Black)
    }

    if (showingBoard) {
        val sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = false)
        ModalBottomSheet(
            onDismissRequest = { showingBoard = false },
            sheetState = sheetState,
            containerColor = AppColors.appSurface,
        ) {
            SportsbookLinesSheet(
                quotes = quotes,
                selectedBookKeys = selectedBookKeys,
                marketTitle = marketTitle,
                selectionTitle = selectionTitle,
                formatLine = formatLine,
            )
        }
    }
}

@Composable
private fun SportsbookLinesSheet(
    quotes: SportsbookMarketQuotes,
    selectedBookKeys: Set<String>,
    marketTitle: String,
    selectionTitle: String,
    formatLine: (Double) -> String,
) {
    val mine = quotes.quotes.filter { it.bookKey in selectedBookKeys }
    val others = quotes.quotes.filter { it.bookKey !in selectedBookKeys }
    Column(Modifier.fillMaxWidth().padding(horizontal = 16.dp, vertical = 8.dp)) {
        Text(marketTitle, color = AppColors.appTextPrimary, fontSize = 18.sp, fontWeight = FontWeight.Black)
        Spacer(Modifier.padding(6.dp))
        LazyColumn(verticalArrangement = Arrangement.spacedBy(0.dp)) {
            if (mine.isNotEmpty()) {
                item { SectionHeader("Your books") }
                items(mine, key = { it.bookKey }) { quote ->
                    QuoteRow(quote, quotes, selectedBookKeys, formatLine)
                }
            }
            item { SectionHeader(if (mine.isEmpty()) selectionTitle else "Everywhere else") }
            items(others, key = { it.bookKey }) { quote ->
                QuoteRow(quote, quotes, selectedBookKeys, formatLine)
            }
            item {
                Text(
                    capturedLabel(quotes.capturedAt),
                    color = AppColors.appTextMuted,
                    fontSize = 11.sp,
                    modifier = Modifier.padding(vertical = 12.dp),
                )
            }
        }
    }
}

@Composable
private fun SectionHeader(title: String) {
    Text(
        title.uppercase(),
        color = AppColors.appTextSecondary,
        fontSize = 11.sp,
        fontWeight = FontWeight.Black,
        modifier = Modifier.padding(top = 10.dp, bottom = 6.dp),
    )
}

@Composable
private fun QuoteRow(
    quote: SportsbookQuote,
    quotes: SportsbookMarketQuotes,
    selectedBookKeys: Set<String>,
    formatLine: (Double) -> String,
) {
    val isBest = quote.bookKey == quotes.best?.bookKey
    val isPinned = quote.bookKey in selectedBookKeys
    Row(
        Modifier.fillMaxWidth().padding(vertical = 8.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(12.dp),
    ) {
        SportsbookLogoView(quote.bookKey, null, SportsbookLogoStyle.COMPACT)
        Column(Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(2.dp)) {
            Text(quote.bookName, color = AppColors.appTextPrimary, fontSize = 14.sp, fontWeight = FontWeight.Bold)
            if (isPinned || isBest) {
                Row(horizontalArrangement = Arrangement.spacedBy(5.dp)) {
                    if (isPinned) Tag("YOURS", AppColors.appAccentBlue)
                    if (isBest) Tag("BEST", AppColors.appPrimary)
                }
            }
        }
        Row(horizontalArrangement = Arrangement.spacedBy(8.dp), verticalAlignment = Alignment.CenterVertically) {
            quote.line?.let {
                Text(
                    formatLine(it),
                    color = if (isBest) AppColors.appPrimary else AppColors.appTextPrimary,
                    fontSize = 15.sp,
                    fontWeight = FontWeight.Black,
                )
            }
            quote.price?.let {
                Text(
                    GameCardFormatting.formatMoneyline(it),
                    color = AppColors.appTextSecondary,
                    fontSize = 13.sp,
                    fontWeight = FontWeight.Bold,
                    fontFamily = FontFamily.Monospace,
                )
            }
        }
    }
}

@Composable
private fun Tag(text: String, tint: Color) {
    Text(
        text,
        color = tint,
        fontSize = 8.sp,
        fontWeight = FontWeight.Black,
        modifier = Modifier
            .clip(RoundedCornerShape(50))
            .background(tint.copy(alpha = 0.14f))
            .padding(horizontal = 5.dp, vertical = 2.dp),
    )
}

private fun priceLabel(quote: SportsbookQuote, formatLine: (Double) -> String): String {
    val line = quote.line?.let(formatLine)
    val price = quote.price?.let { GameCardFormatting.formatMoneyline(it) }
    return listOfNotNull(line, price).joinToString("  ")
}

private fun capturedLabel(capturedAt: String?): String {
    if (capturedAt.isNullOrEmpty()) return "Confirm the number at the book before betting."
    val instant = runCatching { Instant.parse(capturedAt) }.getOrNull()
        ?: return "Confirm the number at the book before betting."
    val minutes = Duration.between(instant, Instant.now()).toMinutes()
    val rel = when {
        minutes < 1 -> "just now"
        minutes < 60 -> "$minutes min ago"
        minutes < 60 * 24 -> "${minutes / 60} hr ago"
        else -> "${minutes / (60 * 24)} days ago"
    }
    return "Lines captured $rel. Confirm at the book before betting."
}

File: wagerproof-ios-native/Wagerproof/Features/GameCards/Components/BestBookChip.swift
import SwiftUI
import WagerproofDesign
import WagerproofModels

/// The sportsbook a card is quoting, as a Liquid Glass pill that opens the full
/// board on tap.
///
/// Sits in the slot the conviction pill used to occupy. A card's job is to say
/// what the bet is and where to get it; "High Conviction" is still shown, but
/// inline with the pick rather than competing with the price for the corner.
///
/// The chip and the board it opens read from the SAME `SportsbookMarketQuotes`,
/// so the number on the pill is always the first row of the list behind it.
struct BestBookChip: View {
    let quotes: SportsbookMarketQuotes
    /// The books the user holds. Empty = no filter, lead with the best number
    /// anywhere.
    let selectedBookKeys: Set<String>
    /// e.g. "Spread" — titles the board.
    let marketTitle: String
    /// e.g. "Patriots +4.5" — what the user is actually being quoted.
    let selectionTitle: String
    /// Renders a spread/total number in the card's own convention so the pill
    /// can't disagree with the bar above it.
    let formatLine: (Double) -> String

    @State private var showingBoard = false

    private var shown: SportsbookQuote? { quotes.preferred(selectedBookKeys) }
    /// False when the user HAS books but none of them price this market, so the
    /// chip is showing the fallback. The label changes to say so.
    private var isFromSelection: Bool { quotes.isFromSelection(selectedBookKeys) }

    var body: some View {
        if let shown {
            Button {
                showingBoard = true
            } label: {
                HStack(spacing: 7) {
                    SportsbookLogoView(
                        logoURL: nil,
                        bookKey: shown.bookKey,
                        bookName: shown.bookName,
                        style: .compact
                    )
                    VStack(alignment: .leading, spacing: 1) {
                        Text(chipLabel)
                            .font(.system(size: 8, weight: .black))
                            .tracking(0.5)
                            .foregroundStyle(Color.appTextMuted)
                        Text(priceLabel(shown))
                            .font(.system(size: 13, weight: .heavy, design: .rounded))
                            .foregroundStyle(Color.appTextPrimary)
                    }
                    Image(systemName: "chevron.down")
                        .font(.system(size: 9, weight: .black))
                        .foregroundStyle(Color.appTextMuted)
                }
                .padding(.horizontal, 10)
                .padding(.vertical, 7)
                .modifier(LiquidGlassCapsule())
            }
            .buttonStyle(.plain)
            .sheet(isPresented: $showingBoard) {
                SportsbookLinesSheet(
                    quotes: quotes,
                    selectedBookKeys: selectedBookKeys,
                    marketTitle: marketTitle,
                    selectionTitle: selectionTitle,
                    formatLine: formatLine
                )
                .presentationDetents([.medium, .large])
                .presentationDragIndicator(.visible)
            }
        }
    }

    /// "YOUR BOOK(S)" only when the number really came from the user's set —
    /// otherwise the chip would credit a book they don't hold.
    private var chipLabel: String {
        guard !selectedBookKeys.isEmpty else { return "BEST LINE" }
        if isFromSelection { return selectedBookKeys.count == 1 ? "YOUR BOOK" : "YOUR BOOKS" }
        return "NOT AT YOUR BOOKS"
    }

    private func priceLabel(_ quote: SportsbookQuote) -> String {
        let line = quote.line.map(formatLine)
        let price = quote.price.map { GameCardFormatting.formatMoneyline($0) }
        return [line, price].compactMap { $0 }.joined(separator: "  ")
    }
}

/// Every book's number on one market, best-first, with the user's own book
/// pinned to the top when they've set one.
struct SportsbookLinesSheet: View {
    let quotes: SportsbookMarketQuotes
    let selectedBookKeys: Set<String>
    let marketTitle: String
    let selectionTitle: String
    let formatLine: (Double) -> String

    @Environment(\.dismiss) private var dismiss

    /// The user's own books lead, still ordered best-first among themselves;
    /// everyone else follows in the same order. A better number elsewhere is
    /// never hidden — just demoted.
    private var mine: [SportsbookQuote] {
        quotes.quotes.filter { selectedBookKeys.contains($0.bookKey) }
    }

    private var others: [SportsbookQuote] {
        quotes.quotes.filter { !selectedBookKeys.contains($0.bookKey) }
    }

    var body: some View {
        NavigationStack {
            List {
                if !mine.isEmpty {
                    Section {
                        ForEach(mine) { quote in row(quote) }
                    } header: {
                        Text("Your books")
                    }
                }
                Section {
                    ForEach(others) { quote in row(quote) }
                } header: {
                    Text(mine.isEmpty ? selectionTitle : "Everywhere else")
                } footer: {
                    if let capturedAt = quotes.capturedAt {
                        // Stated plainly: the capture pauses out of season, and a
                        // board that looks live but is weeks old is worse than one
                        // that admits its age.
                        Text("Lines captured \(capturedAt.formatted(.relative(presentation: .named))). Confirm at the book before betting.")
                    } else {
                        Text("Confirm the number at the book before betting.")
                    }
                }
            }
            .navigationTitle(marketTitle)
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .topBarTrailing) {
                    Button("Done") { dismiss() }
                }
            }
        }
    }

    private func row(_ quote: SportsbookQuote) -> some View {
        let isBest = quote.bookKey == quotes.best?.bookKey
        let isPinned = selectedBookKeys.contains(quote.bookKey)

        return HStack(spacing: 12) {
            SportsbookLogoView(
                logoURL: nil,
                bookKey: quote.bookKey,
                bookName: quote.bookName,
                style: .compact
            )
            VStack(alignment: .leading, spacing: 2) {
                Text(quote.bookName)
                    .font(.system(size: 14, weight: .bold))
                    .foregroundStyle(Color.appTextPrimary)
                if isPinned || isBest {
                    HStack(spacing: 5) {
                        if isPinned { tag("YOURS", tint: .appAccentBlue) }
                        if isBest { tag("BEST", tint: .appPrimary) }
                    }
                }
            }
            Spacer()
            HStack(spacing: 8) {
                if let line = quote.line {
                    Text(formatLine(line))
                        .font(.system(size: 15, weight: .heavy, design: .rounded))
                        .foregroundStyle(isBest ? Color.appPrimary : Color.appTextPrimary)
                }
                if let price = quote.price {
                    Text(GameCardFormatting.formatMoneyline(price))
                        .font(.system(size: 13, weight: .bold, design: .monospaced))
                        .foregroundStyle(Color.appTextSecondary)
                }
            }
        }
        .padding(.vertical, 2)
    }

    private func tag(_ text: String, tint: Color) -> some View {
        Text(text)
            .font(.system(size: 8, weight: .black))
            .tracking(0.4)
            .foregroundStyle(tint)
            .padding(.horizontal, 5)
            .padding(.vertical, 2)
            .background(tint.opacity(0.14), in: Capsule())
    }
}
